src/menu/menu.c

```diff
@@ -86,14 +102,59 @@ item_create(struct menu *menu, const char *text, bool show_arrow)
 	struct server *server = menu->server;
 	struct theme *theme = server->theme;
 
+	const char *arrow = show_arrow ? "›" : NULL;
+
 	if (!menu->item_height) {
 		menu->item_height = font_height(&rc.font_menuitem)
 			+ 2 * MENU_ITEM_PADDING_Y;
 	}
 	menuitem->height = menu->item_height;
 
 	int x, y;
-	int item_max_width = MENUWIDTH - 2 * MENU_ITEM_PADDING_X;
+	int text_width = font_width(&rc.font_menuitem, text);
+	if (arrow) {
+		text_width += font_width(&rc.font_menuitem, arrow);
+	}
+	text_width = clamp(text_width, theme->menu_min_width, theme->menu_max_width);
+
+	if (menu->size.width >= text_width + 2 * MENU_ITEM_PADDING_X) {
+		/* Reuse existing menu width */
+		text_width = menu->size.width - 2 * MENU_ITEM_PADDING_X;
+	} else {
+		/* Store new menu width and update existing items */
```

Comment:
Guess this could actually be split into its own static function.
Something like `menu_update_width(struct menu *menu, int new_width)`.

Reply:
Are we updating the width of the menu and all its items here on every `item_create()`?
Would it make more sense to calculate and set/update the menu width once on hitting `</menu>`?
I think that would mean calling such a function here: https://github.com/labwc/labwc/blob/aa4ef94af02c9488f0f8e2be25912ee4b358c4ff/src/menu/menu.c#L402

Reply:
I thought about that but the issue was that to get the max font width of all menu items within a menu we need the actual text (to call `font_width()`). But we only store the `char *` within the `scaled_font_buffer` and its not really intended to access that from the outside. And storing another copy of the `char *` within each item feels redundant.

Maybe `item_create()` could store the calculated font_width within the `struct menuitem` though, that would just be an additional `int` per item. Then the resizing could indeed be done just once when encountering a `</menu>`.

Reply:
Sorry - the rightful place would probably be here  after traverse()...
https://github.com/labwc/labwc/blob/aa4ef94af02c9488f0f8e2be25912ee4b358c4ff/src/menu/menu.c#L360
Not tested - just reading the code.

Reply:
> I thought about that but the issue was that to get the max font width of all menu items within a menu we need the actual text (to call `font_width()`). But we only store the `char *` within the `scaled_font_buffer` and its not really intended to get accessed from the outside. Storing another copy of the `char *` within each item feels redundant.

Sorry - yes, have just seen your comment in OP :smile:

> Maybe `item_create()` could store the calculated font_width within the `struct menuitem` though, that would just be an additional `int` per item. Then the resizing could indeed be done just once when encountering a `</menu>`.

Good idea. That seems the cleanest.

Reply:
Did that and did some minor refactoring to only expose a `menu_init()` which then calls `init_rootmenu()` and `init_windowmenu()` internally (which were changed to being `static`). Then added a `post_processing()` step there which triggers the menu width calculations / resizing for all menus.